### ISSUE TITLE
Mechanism for working with vision-based models

### ISSUE THREAD

**Issue by daedsidog** [karthink/gptel#244]
A lot of times I need to pass information to ChatGPT that I can't copy, such as a snippet from an old, scanned document or formatted mathematics.

Right now I have a manual process where I query a visual model (though a website) so that he tells me what he sees. E.g., when I paste him a snippet of mathematics, he will give me LaTeX code, which I then pass down to ChatGPT.

Would be very nice to have something like this.

**Comment by karthink** [comment]
I forget, do you use org-mode or markdown with gptel?

**Comment by daedsidog** [comment]
I used to use Org mode but I switched to Markdown because I was tired of gptel sometimes doing weird things (like removing underscores). I think it's fixed in the latest version, but I haven't switched.

Why is that relevant, though?

**Comment by karthink** [comment]
> I used to use Org mode but I switched to Markdown because I was tired of gptel sometimes doing weird things (like removing underscores). I think it's fixed in the latest version, but I haven't switched back.

It should be fixed now, yeah.

> Why is that relevant, though?

It's easier to support vision models in Org mode.  That said, please see the discussion in #231.

**Comment by daedsidog** [comment]
> It's easier to support vision models in Org mode.  That said, please see the discussion in [#231](https://github.com/karthink/gptel/discussions/231).

Interesting.

I honestly think the most power from gptel comes with just the abstraction layer it provides when interacting with various models. I, for one, have completely eliminated the process of manually typing code by implementing context generation. Below is a demonstration of me constructing a context buffer, and with a keypress I use gptel's replace-in-place with it. Works exceedingly well. This is kind of its own separate thing from gptel, but I was wondering if the scope of gptel should include this sort of thing.

![contexter](https://github.com/karthink/gptel/assets/41439659/aca82b11-2b97-46e1-88a5-177653316059)

What I want now is just a way, totally unrelated to Org mode or MD, which will allow me to "send" ChatGPT queries with images (i.e., send current image saved on clip) and get an input in place.

**Comment by karthink** [comment]
>  I was wondering if the scope of gptel should include this sort of thing.

I'm interested to understand what you mean here -- I just had trouble following the demo.

**Comment by daedsidog** [comment]
> > I was wondering if the scope of gptel should include this sort of thing.
> 
> I'm interested to understand what you mean here -- I just had trouble following the demo.

My apologies, my explanation was terrible.

The demo showcases a way to mark areas in different buffers, and aggregate them in their own dedicated buffer. That buffer can then be copied and handed to gptel as context. This is much easier than manually copy pasting sections of context into the dedicated chat buffer/external ChatGPT website, and also has the added bonus of minimizing the context by collapsing code that doesn't contribute to the context.

You can manually remove context snippets from the context buffer.

In a nutshell, it's a glorified yanker, but I found it incredibly useful.

> I'm interested to understand what you mean here

I am wondering if you would be open for this to be integrated into gptel, or should this remain its own separate package. It's pretty useless outside of gptel, though.

**Comment by karthink** [comment]
> I am wondering if you would be open for this to be integrated into gptel, or should this remain its own separate package. It's pretty useless outside of gptel, though.

I like the idea!  I'll have to think about how to integrate it into gptel though.  Right now the best idea I have is "Add an option to the transient menu to append a selected region to the system prompt".  You've developed a more sophisticated UI for this style of usage.

**Comment by karthink** [comment]
Converting to a discussion since there's nothing to fix in gptel right now.
